From 7e14606156e365f3498b614aaa7d992eaa1bb6f6 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Sun, 28 Oct 2018 11:32:33 +0200
Subject: [PATCH 0198/1239] arm64: mvebu: a3900: fix the USB 2.0 TX amplitude

Fix the default settings for USB 2.0 (UTMI) ports TX output
driver amplitude value for passing compatibility tests.

Change-Id: Ic5babec8cee65d6789ea2ea129fa7e658375eb5a
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
---
 board/Marvell/mvebu_armada-8k/board.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/board/Marvell/mvebu_armada-8k/board.c b/board/Marvell/mvebu_armada-8k/board.c
index aa6b9126f1..3fd8936203 100644
--- a/board/Marvell/mvebu_armada-8k/board.c
+++ b/board/Marvell/mvebu_armada-8k/board.c
@@ -14,6 +14,12 @@
 #include <mvebu/cfg_eeprom.h>
 #endif
 
+#define CP_USB20_BASE_REG(cp, p)	(MVEBU_REGS_BASE_CP(0, cp) + \
+						0x00580000 + 0x1000 * (p))
+#define CP_USB20_TX_CTRL_REG(cp, p)	(CP_USB20_BASE_REG(cp, p) + 0xC)
+#define CP_USB20_TX_OUT_AMPL_MASK	(0x7 << 20)
+#define CP_USB20_TX_OUT_AMPL_VALUE	(0x3 << 20)
+
 DECLARE_GLOBAL_DATA_PTR;
 
 int __soc_early_init_f(void)
@@ -42,7 +48,7 @@ int board_early_init_r(void)
 
 int board_init(void)
 {
-	/* adress of boot parameters */
+	/* address of boot parameters */
 	gd->bd->bi_boot_params = CONFIG_SYS_SDRAM_BASE + 0x100;
 
 #ifdef CONFIG_BOARD_CONFIG_EEPROM
@@ -56,5 +62,17 @@ int board_late_init(void)
 {
 	/* Pre-configure the USB ports (overcurrent, VBus) */
 
+	/* Adjust the USB 2.0 port TX output driver amplitude
+	 * for passing compatibility tests
+	 */
+	if (of_machine_is_compatible("marvell,armada3900-vd")) {
+		u32 port;
+
+		for (port = 0; port < 2; port++)
+			clrsetbits_le32(CP_USB20_TX_CTRL_REG(0, port),
+					CP_USB20_TX_OUT_AMPL_MASK,
+					CP_USB20_TX_OUT_AMPL_VALUE);
+	}
+
 	return 0;
 }
-- 
2.29.0

